src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
+            assert(FALSE && "Couldn't open the search iterator.");
+            return -1;
+        }
+
+        UStringSearch* pNull = NULL;
+        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], *pSearchIterator, pNull))
+        {
+            usearch_close(*pSearchIterator);
+            *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+        }
+    }
+
+    assert(*pSearchIterator != NULL && "pSearch not expected to be null here.");
+
+    if (*pSearchIterator == USED_STRING_SEARCH || !pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], USED_STRING_SEARCH, *pSearchIterator))
+    {
+        // The handle is in use. Create a temporary new one.
+        *pSearchIterator = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, pErr);
+        return -1;
+    }
+
+    usearch_setText(*pSearchIterator, lpSource, cwSourceLength, pErr);
+    if (!U_SUCCESS(*pErr))
+    {
+        // restore the handle back to the cache.
+        pSortHandle->searchIteratorPerOption[options] = *pSearchIterator;
+        return -1;
+    }
+
+    usearch_setPattern(*pSearchIterator, lpTarget, cwTargetLength, pErr);
+    if (!U_SUCCESS(*pErr))
+    {
+        // restore the handle back to the cache.
+        pSortHandle->searchIteratorPerOption[options] = *pSearchIterator;
+        return -1;
+    }
+
+    return options;
+}
+
+// If succeeded, U_SUCCESS(*pErr) should be true.
+// return -1 if couldn't borrow search handle from the SortHandle cache, otherwise, it return the slot number of the cache.
+static inline const int32_t GetSearchIteratorFromSortHandle(
+                        SortHandle* pSortHandle,
+                        const UChar* lpTarget,
+                        int32_t cwTargetLength,
+                        const UChar* lpSource,
+                        int32_t cwSourceLength,
+                        int32_t options,
+                        UErrorCode* pErr,
+                        UStringSearch** pSearchIterator)
+{
+    const UCollator* pColl = GetCollatorFromSortHandle(pSortHandle, options, pErr);
+    if (!U_SUCCESS(*pErr))
+    {
+        assert(FALSE && "Couldn't get the collator.");
+        return -1;
+    }
+
+    return GetSearchIteratorFromSortHandleUsingCollator(
+                        pSortHandle,
+                        pColl,
+                        lpTarget,
+                        cwTargetLength,
+                        lpSource,
+                        cwSourceLength,
+                        options,
+                        pErr,
+                        pSearchIterator);
+}
+
+static inline void  ReleaseSearchHandle(SortHandle* pSortHandle, UStringSearch* pSearchIterator, int32_t searchCacheSlot)
+{
+    if (searchCacheSlot < 0)
+    {
+        usearch_close(pSearchIterator);
+    }
+    else
+    {
+        pSortHandle->searchIteratorPerOption[searchCacheSlot] = pSearchIterator;
```

Comment:
I wonder what happens with the reference to `text` and `pattern` set via `usearch_setText` and `usearch_setPattern`in `GetSearchIteratorFromSortHandleUsingCollator`.

Are we risking a memory leak? Should the `text` and `pattern` be somehow set to null here?

Reply:
I don't think there is any risk here. GC has no knowledge about the native pointers is assigned in some cache. GC doesn't track any native pointer operation on the string. The string can be collected after the call return and the object is not reachable.